kolibri/plugins/document_epub_render/assets/src/views/TableWrap.js

```diff
@@ -0,0 +1,231 @@
+/* TODO: Get rid of this file when Firefox breaks tables properly in CSS column layout
+ * Known issues: https://caniuse.com/#feat=multicolumn
+ */
+const DEFAULT_PADDING = 40;
+const Factory = {
+  /**
+   * @param {HTMLElement} el
+   * @returns {Table}
+   */
+  buildTable(el) {
+    return new Table(el);
+  },
+
+  /**
+   * @param {Document} document
+   * @returns {FixTableWrap}
+   */
+  buildFixer(document) {
+    return new FixTableWrap(document);
+  },
+};
+
+export default Factory;
+
+/**
+ * @param {HTMLElement} element
+ * @param {Number} [defaultHeight]
+ * @returns {Number}
+ */
+export const getInnerHeight = function(element, defaultHeight) {
+  defaultHeight = defaultHeight || element.clientHeight;
+
+  try {
+    const styles = global.getComputedStyle(element);
+    const height = parseInt(styles.getPropertyValue('height')) || defaultHeight;
+
+    if (styles.getPropertyValue('box-sizing') !== 'border-box') {
+      return height;
+    }
+
+    const props = ['padding-top', 'padding-bottom', 'border-top-width', 'border-bottom-width'];
+
+    return props
+      .map(prop => parseInt(styles.getPropertyValue(prop)) || 0)
+      .reduce((height, style) => height - style, height);
+  } catch (e) {
+    return defaultHeight;
+  }
+};
+
+/**
+ * Splits tables in one or more pieces if the tables are overflowing in what is assumed to be a
+ * CSS column layout, where they would otherwise be wrapping from column to column
+ */
+export class FixTableWrap {
+  /**
+   * @param {Document} document
+   * @param {Object} [factory]
+   * @param {Function} factory.buildTable
+   */
+  constructor(document, factory) {
+    this._document = document;
+    this._maxHeight = getInnerHeight(document.body, document.body.clientHeight - DEFAULT_PADDING);
+    this._factory = factory || Factory;
+  }
+
+  /**
+   * Fix all tables in document that do not fit within their container's height
+   */
+  fix() {
+    this.getTablesToFix().forEach(table => {
+      const container = table.get().parentElement;
+
+      table.split(this._maxHeight).forEach(newTable => {
+        container.insertBefore(newTable.get(), table.get());
+
+        if (!newTable.check(this._maxHeight)) {
+          this.makeScrollable(newTable);
+        }
+      });
+
+      container.removeChild(table.get());
+    });
+  }
+
+  /**
+   * Makes table "scrollable" by making the cells scrollable. Since the tables are
+   * split by rows, then this table has one row, so we'll scroll the cells.
+   *
+   * @param {Table} table
+   */
+  makeScrollable(table) {
```

Comment:
This might need to be called every time the font-size is changed. I see that it's only called when the epub loads initially.

Reply:
For me it appears to be called every time for font sizing increasing / decreasing.

Reply:
Hmm, maybe it isn't. Let me see.